### PR TITLE
[SSO] Add SsoBackend for authenticating users after a successful SAML2 handshake

### PR DESCRIPTION
**Pull request by biyeun**
## Summary
This adds a new authentication backend `SsoBackend` that is used to authenticate a user after a successful SAML2 handshake. 

## Feature Flag
`ENTERPRISE_SSO`

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
New tests included for `SsoBackend`

### QA Plan
Like with all SSO work, there will be a major QA and pen test once the dev work has reached a more complete stage.

### Safety story
Note that this will not be active/accessible on production as there are no active `IdentityProvider`s, so this is 100% safe to merge without QA. 

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/sso/tests/test_backends.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        SsoBackend should fail to move past the first check because an idp_slug
+         was not passed to authenticate()
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username
+        )
+        self.assertIsNone(user)
+        with self.assertRaises(AttributeError):
```
Comment (mjriley) [inline corehq/apps/sso/tests/test_backends.py:69]: Curious about the semantics here. If I don't specify an identity provider slug, why don't I receive an error message like the other failures?

File: corehq/apps/sso/tests/test_backends.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        """
+        SsoBackend should fail to return a user if the Identity Provider
+        slug associated with that user does not exist. It should also populate
+        request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username=self.user.username,
+            idp_slug='doesnotexist'
```
Comment (mjriley) [inline corehq/apps/sso/tests/test_backends.py:94]: I appreciate the explicit strings here like `doesnotexist` and `badusername`

File: corehq/apps/sso/tests/test_backends.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        error message.
+
+        todo this test will change with additional user creation workflows
+         that will be introduced later
+        """
+        self._set_saml_session_index()
+        AuthenticatedEmailDomain.objects.create(
+            email_domain='vaultwax.com',
+            identity_provider=self.idp,
+        )
```
Comment (mjriley) [inline corehq/apps/sso/tests/test_backends.py:196]: Would it make sense to move this into the setup? Like, this suite has a given domain that is always covered by a specific IDP, and then the tests that want to change that assumption verify against a different domain or provider? This feels like an unrelated implementation detail in this specific test.

File: corehq/apps/sso/tests/test_backends.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        SsoBackend should fail to return a user if the username passed to it
+        is not a valid email address (missing email domain / no `@`). It should
+        also populate request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username='badusername',
+            idp_slug=self.idp.slug
+        )
```
Comment (mjriley) [inline corehq/apps/sso/tests/test_backends.py:133]: Just a suggestion, but I've liked creating helper methods for situations like this, where only one of the parameters is relevant to the test (`username='badusername'`). Creating sensible defaults for all of the parameters then allows you to just do something like: `self.authenticate(username='badusername')`

File: corehq/apps/sso/views/saml.py [hunk 5]
```diff
@@ -92,6 +98,7 @@ def sso_saml_acs(request, idp_slug):
         "processed_response": processed_response,
         "saml_relay": saml_relay,
         "request_session_data": request_session_data,
+        "login_error": getattr(request, 'sso_login_error', None),
```
Comment (mjriley) [inline corehq/apps/sso/views/saml.py:101]: Is this response for debugging purposes? If not, are we exposing any potentially sensitive information here? (Moreso with the other properties here like `request_session_data` -- `sso_login_error` doesn't expose anything)

**Review by mjriley — APPROVED** [review]
Looks good. Nothing blocking, most of my feedback was related to the tests.

File: corehq/apps/sso/views/saml.py (hunk 5 above)
Comment (biyeun) [reply corehq/apps/sso/views/saml.py:101]: yeah this whole response is debug only. there will be a final redirect to a page with more user-friendly responses

File: corehq/apps/sso/backends.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        try:
+            user = User.objects.get(username=username)
+        except User.DoesNotExist:
+            # todo handle user creation based on information from request/session
+            #  do this prior to handling the invite scenario and new user scenario
+            request.sso_login_error = f"User {username} does not exist."
+            return None
+
+        request.sso_login_error = None
+        # todo what happens with 2FA required here?
```
Comment (millerdev) [inline corehq/apps/sso/backends.py:56]: Are you expecting to address this comment later or was it intended to be part of this PR?

File: corehq/apps/sso/tests/test_backends.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        cls.user.delete(None)
+        cls.domain.delete()
+        cls.account.delete()
+        super().tearDownClass()
+
+    def setUp(self):
+        super().setUp()
+        # Identity Provider is_active is always True by default for each test
+        self.idp.is_active = True
+        self.idp.save()
```
Comment (millerdev) [inline corehq/apps/sso/tests/test_backends.py:46]: Do tests cause this to be reset to false? If not, could it be moved into `setUpClass`?

File: corehq/apps/sso/backends.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None, **kwargs):
+        #  Note: when implementing ODIC or other protocols in the future,
+        #   a different request.session check can be made
+        if not (username and idp_slug and request.session.get('samlSessionIndex')):
```
Comment (millerdev) [inline corehq/apps/sso/backends.py:15]: Is the presence of `samlSessionIndex` in the session indicative of a successful SAML2 handshake?

File: corehq/apps/sso/backends.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            return None
+
+        try:
+            identity_provider = IdentityProvider.objects.get(slug=idp_slug)
+        except IdentityProvider.DoesNotExist:
+            # not sure how we would even get here, but just in case
+            request.sso_login_error = "Identity Provider does not exist."
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/backends.py:22]: no these are for debugging purposes on our end and will not be shown to the user

File: corehq/apps/sso/backends.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+        try:
+            user = User.objects.get(username=username)
+        except User.DoesNotExist:
+            # todo handle user creation based on information from request/session
+            #  do this prior to handling the invite scenario and new user scenario
+            request.sso_login_error = f"User {username} does not exist."
+            return None
+
+        request.sso_login_error = None
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/backends.py:55]: django eagerly tries each backend in order when `authenticate` is called. if the backend fails because sso-specific stuff is missing, it's not necessarily an `sso_login_error` but that this backend was not appropriate based on the `kwargs` given.

File: corehq/apps/sso/tests/test_backends.py (hunk 1 above)
Comment (biyeun) [reply corehq/apps/sso/tests/test_backends.py:69]: I mentioned this in another comment, but it's because that backend was not appropriate for the given set of `kwargs`. django will try each authentication backend in order when `authenticate` is called. If a slug wasn't present, it means `authenticate` was called somewhere else and it was unrelated to sso.

File: corehq/apps/sso/backends.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/backends.py:12]: won't removing the `kwarg` defaults result in a 500 when django loops through all the backends trying to match one with an `authenticate` request?

File: corehq/apps/sso/tests/test_backends.py (hunk 7 above)
Comment (biyeun) [reply corehq/apps/sso/tests/test_backends.py:46]: some tests set this to `False` which is why it's not in `setUpClass`

File: corehq/apps/sso/backends.py (hunk 8 above)
Comment (biyeun) [reply corehq/apps/sso/backends.py:15]: yes. And this is something I'm curious to open discussion about. `authenticate` using these specific `kwargs` would only be called with in the saml `ACS` view after the handshake is successful, and nowhere else. But should the `samlSessionIndex` be stored in postgres temporarily so that it can be cross-referenced here? My understanding is that unless someone has access to the `SECRET_KEY` the contents in `session` can't be spoofed. On top of it since the only trigger with all the required `kwargs` comes from within the `acs` that by itself adds security. no? am I incorrect in this understanding?

File: corehq/apps/sso/backends.py (hunk 6 above)
Comment (biyeun) [reply corehq/apps/sso/backends.py:56]: intentionally left as a todo for later :)

File: corehq/apps/sso/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/sso/backends.py:12]: Oh, I didn't know this will be called by code we did not write. Fine to keep the defaults if the interface it must provide includes being called without those args.

File: corehq/apps/sso/backends.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/backends.py:22]: shouldn't be a security concern as in the end this message will not be shown to the user

File: corehq/apps/sso/tests/test_backends.py (hunk 7 above)
Comment (millerdev) [reply corehq/apps/sso/tests/test_backends.py:46]: Looks like only one test does that. What do you think of using `addCleanup` in that test? Otherwise it's an extra database hit for every test.

File: corehq/apps/sso/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/sso/backends.py:12]: Another question related to https://github.com/dimagi/commcare-hq/pull/29215#discussion_r581345588, will Django ever call this with `username` and `idp_slug`? If not, then why are we exposing it to Django as an authentication backend (since authentication will always fail when Django calls it)?

File: corehq/apps/sso/backends.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/sso/backends.py:55]: I see. I think it would be nice for that to be made more explicit, i.e:
```
if not self.should_service_request(params):
  return None

# initialize values
# handle error conditions
# return
```

File: corehq/apps/sso/tests/test_backends.py (hunk 7 above)
Comment (biyeun) [reply corehq/apps/sso/tests/test_backends.py:46]: i think i might just add this to the one test that sets it to false

File: corehq/apps/sso/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/sso/backends.py:12]: that's a good question. I think I was just trying to follow the django pattern for authentication. If I should pull this out entirely of being a django-style backend, then I should just do that I suppose...thoughts? Then there would be absolutely zero chance of this being called outside of saml's `acs` view

[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/sso/backends.py:12]: Yes, that seems like the right thing to do.

**Comment by biyeun** [comment]
@millerdev alright. I ended up refactoring basically everything and moved from using the django pattern for backends to creating a method specifically for fetching a user only from the SAML acs view (or any successful sso auth view in the future).

File: corehq/apps/sso/authentication.py [hunk 12]
```diff
@@ -1,57 +1,44 @@
...
+        )
+
+    try:
+        user = User.objects.get(username=username)
+    except User.DoesNotExist:
+        # todo handle user creation based on information from request/session
+        #  do this prior to handling the invite scenario and new user scenario
+        raise SsoAuthenticationError(f"User {username} does not exist.")
+
+    # todo what happens with 2FA required here?
```
Comment (millerdev) [inline corehq/apps/sso/authentication.py:43]: Is this comment still valid? Seems like 2FA is not a concern of this function.

File: corehq/apps/sso/tests/test_authentication.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        self.idp.is_active = False
+        self.idp.save()
+        with self.assertRaises(SsoAuthenticationError):
+            get_authenticated_sso_user(
+                self.user.username,
+                self.idp.slug
+            )
+
+        self.idp.is_active = True
+        self.idp.save()
```
Comment (millerdev) [inline corehq/apps/sso/tests/test_authentication.py:76]: This should be done with `self.addCleanup(...)` or `try`/`finally`. Otherwise if this test fails (before this code is executed) it will cause other tests to fail as well. It's easiest to do try/finally in a suggestion on github, but feel free to use `addCleanup` instead if you prefer that.
```suggestion
        try:
            with self.assertRaises(SsoAuthenticationError):
                get_authenticated_sso_user(
                    self.user.username,
                    self.idp.slug
                )
        finally:
            self.idp.is_active = True
            self.idp.save()
```

**Review by millerdev — APPROVED** [review]
I think this arrangement is much better.

**Comment by biyeun** [comment]
sorry for the rebase and force-push here. @millerdev and I offlined yesterday and came to the conclusion that, yes, we have to follow the django pattern and have an auth backend. we also verified that, yes, it is safe due to some django 'magic' that checks whether any of the specified backends' method signatures match the call to `auth.authenticate()`

As a result, I had to backpedal and hard reset to the point prior to the mini-refactor.

File: corehq/apps/sso/backends.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None,
+                     is_handshake_successful=False, **kwargs):
```
Comment (millerdev) [inline corehq/apps/sso/backends.py:13]: These argument defaults are not necessary
```suggestion
    def authenticate(self, request, username, idp_slug, is_handshake_successful):
```

Django's [`auth.authenticate` function](https://github.com/django/django/blob/stable/2.2.x/django/contrib/auth/__init__.py#L67-L71) will not try to call this method if the supplied `**credentials` do not match the arguments required here. In fact, it is more efficient to make them required (and not accept `**kwargs`) because it will cause the method to _not_ be called on other non-SSO authentication attempts. As an example, note how [`RemoteUserBackend`](https://github.com/django/django/blob/stable/2.2.x/django/contrib/auth/backends.py#L128) is implemented with no defaults, and it also extends `ModelBackend`.

File: corehq/apps/sso/backends.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        if not (username and idp_slug and is_handshake_successful):
```
Comment (millerdev) [inline corehq/apps/sso/backends.py:14]: Should probably also check `request` here since it could theoretically be `None`:
> request is an `HttpRequest` and may be `None` if it wasn’t provided to `authenticate()` (which passes it on to the backend).

https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend

In practice this is currently not super important since we always supply a `request` when calling `auth.authenticate()` in a way that will match this backend. However, it is more future proof to do the check.

**Review by millerdev — APPROVED** [review]

**Review by millerdev — APPROVED** [review]
